launcher/src/main/java/org/apache/spark/launcher/AbstractLauncher.java

```diff
@@ -0,0 +1,600 @@
+    }
+
+    String sparkHome = getSparkHome();
+    File scala210 = new File(sparkHome, "assembly/target/scala-2.10");
+    File scala211 = new File(sparkHome, "assembly/target/scala-2.11");
+    if (scala210.isDirectory() && scala211.isDirectory()) {
+      checkState(false,
+        "Presence of build for both scala versions (2.10 and 2.11) detected.\n" +
+        "Either clean one of them or set SPARK_SCALA_VERSION in your environment.");
+    } else if (scala210.isDirectory()) {
+      return "2.10";
+    } else {
+      checkState(scala211.isDirectory(), "Cannot find any assembly build directories.");
+      return "2.11";
+    }
+
+    throw new IllegalStateException("Should not reach here.");
+  }
+
+  /**
+   * Which OS is running defines two things:
+   * - the name of the environment variable used to define the lookup path for native libs
+   * - how to execute the command in general.
+   *
+   * The name is easy: PATH on Win32, DYLD_LIBRARY_PATH on MacOS, LD_LIBRARY_PATH elsewhere.
+   *
+   * On Unix-like, we're assuming bash is available. So we print one argument per line to
+   * the output, and use bash's array handling to execute the right thing.
+   *
+    * For Win32, see {@link #prepareForWindows(List<String>,String)}.
+   */
+  protected List<String> prepareForOs(List<String> cmd,
+      String libPath,
+      Map<String, String> env) {
+
+    // If SPARK_HOME does not come from the environment, explicitly set it
+    // in the child's environment.
+    Map<String, String> childEnv = env;
+    if (System.getenv("SPARK_HOME") == null && !env.containsKey("SPARK_HOME")) {
+      childEnv = new HashMap<String, String>(env);
+      childEnv.put("SPARK_HOME", sparkHome);
+    }
+
+    if (isWindows()) {
+      return prepareForWindows(cmd, libPath, childEnv);
+    }
+
+    if (isEmpty(libPath) && childEnv.isEmpty()) {
+      return cmd;
+    }
+
+    List<String> newCmd = new ArrayList<String>();
+    newCmd.add("env");
+
+    if (!isEmpty(libPath)) {
+      String envName = getLibPathEnvName();
+      String currEnvValue = getenv(envName);
+      String newEnvValue = join(File.pathSeparator, currEnvValue, libPath);
+      newCmd.add(String.format("%s=%s", envName, newEnvValue));
+    }
+    for (Map.Entry<String, String> e : childEnv.entrySet()) {
+      newCmd.add(String.format("%s=%s", e.getKey(), e.getValue()));
+    }
+    newCmd.addAll(cmd);
+    return newCmd;
+  }
+
+  protected String shQuote(String s) {
+    StringBuilder quoted = new StringBuilder();
+    boolean hasWhitespace = false;
+    for (int i = 0; i < s.length(); i++) {
+      if (Character.isWhitespace(s.codePointAt(i))) {
+        quoted.append('"');
+        hasWhitespace = true;
+        break;
+      }
+    }
+
+    for (int i = 0; i < s.length(); i++) {
+      int cp = s.codePointAt(i);
+      switch (cp) {
+        case '\'':
+          if (hasWhitespace) {
+            quoted.appendCodePoint(cp);
+            break;
+          }
+        case '"':
```

Comment:
if you have a double quote here that falls through, don't you end up having something like
`key="value contains a " double quote"`?